File: Compiler/src/compiler/operators/Compare.java
package compiler.operators;

import compiler.Operator;
import util.CompilationResult;
import util.DataType;
import util.JenaUtil;
import util.NamingManager;

import java.util.ArrayList;
import java.util.List;

public class Compare extends BaseOperator {

    /**
     * Конструктор
     * @param args Аргументы
     */
    public Compare(List<Operator> args) {
        super(args);
    }

    @Override
    public List<List<DataType>> argsDataTypes() {
        List<List<DataType>> result = new ArrayList<>();

        result.add(List.of(DataType.INTEGER, DataType.DOUBLE));
        result.add(List.of(DataType.DOUBLE, DataType.INTEGER));
        result.add(List.of(DataType.INTEGER, DataType.INTEGER));
        result.add(List.of(DataType.DOUBLE, DataType.DOUBLE));
        result.add(List.of(DataType.STRING, DataType.STRING));

        return result;
    }

    @Override
    public DataType resultDataType() {
        return DataType.COMPARISON_RESULT;
    }

    @Override
    public CompilationResult compile() {
        // TODO: флаг "не может быть неопределенным" позволит избежать паузы

        // Объявляем переменные
        String value = "";
        String ruleHead = "";
        String completedRules = "";

        // Получаем аргументы
        Operator arg0 = arg(0);
        Operator arg1 = arg(1);

        // Компилируем аргументы
        CompilationResult compiledArg0 = arg0.compile();
        CompilationResult compiledArg1 = arg1.compile();

        // Генерируем имена
        String skolemName = NamingManager.genVarName();
        String resFlagName = NamingManager.genPredName();

        // Подставляем в сколем все объекты операторов
        StringBuilder skolemArgs = new StringBuilder().append(skolemName);
        for(String val : arg0.objectsUsedInRule()) {
            skolemArgs.append(",");
            skolemArgs.append(val);
        }
        for(String val : arg1.objectsUsedInRule()) {
            skolemArgs.append(",");
            skolemArgs.append(val);
        }

        // Правило для проверки эквивалентности
        String operatorsPart = compiledArg0.ruleHead() + compiledArg1.ruleHead();
        String equal = operatorsPart + JenaUtil.genEqualPrim(compiledArg0.value(), compiledArg1.value());
        String equalRule = JenaUtil.genRule(
                equal,
                skolemArgs.toString(),
                skolemName,
                resFlagName,
                JenaUtil.genStingVal("equal"));

        // Собираем правила
        completedRules = compiledArg0.completedRules() + compiledArg1.completedRules() + equalRule;

        // Правила для проверки больше/меньше (только для чисел)
        if(arg0.resultDataType() != DataType.STRING) {
            // Правило для проверки меньше
            String less = operatorsPart + JenaUtil.genLessThanPrim(compiledArg0.value(), compiledArg1.value());
            String lessRule = JenaUtil.genRule(
                    less,
                    skolemArgs.toString(),
                    skolemName,
                    resFlagName,
                    JenaUtil.genStingVal("less"));

            // Правило для проверки больше
            String greater = operatorsPart + JenaUtil.genGreaterThanPrim(compiledArg0.value(), compiledArg1.value());
            String greaterRule = JenaUtil.genRule(
                    greater,
                    skolemArgs.toString(),
                    skolemName,
                    resFlagName,
                    JenaUtil.genStingVal("greater"));

            completedRules += lessRule + greaterRule;
        }

        completedRules += JenaUtil.PAUSE_MARK;

        String empty = NamingManager.genVarName();

        // Правило для проверки undetermined
        String undetermined = JenaUtil.genNotEqualPrim(empty, resFlagName);
        String undeterminedRule = JenaUtil.genRule(
                undetermined,
                skolemArgs.toString(),
                skolemName,
                resFlagName,
                JenaUtil.genStingVal("undetermined"));

        completedRules += undeterminedRule;

        value = NamingManager.genVarName();

        // FIXME?: придумать способ инициализации получше
        // Инициализируем переменные в новом правиле
        StringBuilder initPart = new StringBuilder();
        for(String var : arg0.objectsUsedInRule()) {
            initPart.append(JenaUtil.genTriple(var, NamingManager.genVarName(), NamingManager.genVarName()));
        }
        for(String var : arg1.objectsUsedInRule()) {
            initPart.append(JenaUtil.genTriple(var, NamingManager.genVarName(), NamingManager.genVarName()));
        }
        if(arg0.objectsUsedInRule().isEmpty() && arg1.objectsUsedInRule().isEmpty()) {
            initPart.append(JenaUtil.genTriple(NamingManager.genVarName(), NamingManager.genVarName(), NamingManager.genVarName()));
        }

        ruleHead = initPart + JenaUtil.genMakeSkolemPrim(skolemArgs.toString()) +
                JenaUtil.genTriple(skolemName, resFlagName, value);

        usedObjects = new ArrayList<>(arg0.objectsUsedInRule());
        usedObjects.addAll(new ArrayList<>(arg1.objectsUsedInRule()));

        return new CompilationResult(value, ruleHead, completedRules);
    }
}

File: Compiler/src/compiler/operators/Compare.kt
package compiler.operators

import compiler.Operator
import compiler.util.CompilationResult
import util.DataType
import util.JenaUtil.COMPARE_RESULT_PREDICATE
import util.JenaUtil.POAS_PREF
import util.JenaUtil.genEqualPrim
import util.JenaUtil.genGreaterThanPrim
import util.JenaUtil.genLessThanPrim
import util.JenaUtil.genLink
import util.JenaUtil.genNotEqualPrim
import util.JenaUtil.genTriple
import util.JenaUtil.genVal
import util.NamingManager.genVarName

/**
 * Сравнение
 * TODO?: сравнение объектов на больше/меньше?
 */
class Compare(args: List<Operator>) : BaseOperator(args) {

    override val argsDataTypes
        get() = listOf(
            listOf(DataType.Integer, DataType.Double),
            listOf(DataType.Double, DataType.Integer),
            listOf(DataType.Integer, DataType.Integer),
            listOf(DataType.Double, DataType.Double),
            listOf(DataType.String, DataType.String),
            listOf(DataType.Object, DataType.Object)
        )

    override val resultDataType get() = DataType.ComparisonResult

    override fun compile(): CompilationResult {
        // Объявляем переменные
        val value = genVarName()
        val heads = ArrayList<String>()
        var completedRules = ""

        // Получаем аргументы
        val arg0 = arg(0)
        val arg1 = arg(1)

        // Компилируем аргументы
        val compiledArg0 = arg0.compile()
        val compiledArg1 = arg1.compile()

        // Передаем завершенные правила дальше
        completedRules += compiledArg0.rules +
                compiledArg1.rules

        // Вспомогательные переменные
        val empty = genVarName()
        val compareResultPredicate = genLink(POAS_PREF, COMPARE_RESULT_PREDICATE)

        // Для всех результатов компиляции
        compiledArg0.bodies.forEach { head0 ->
            compiledArg1.bodies.forEach { head1 ->
                // Если аргументы можно сравнивать только на эквивалентность
                if (arg0.resultDataType == DataType.Object
                    || arg0.resultDataType == DataType.String
                ) {
                    // Правило для эквивалентности
                    var equalHead = head0 + head1
                    equalHead += genEqualPrim(compiledArg0.value, compiledArg1.value)
                    // Инициализация value
                    equalHead += genTriple(
                        empty,
                        compareResultPredicate,
                        genVal(compiler.util.ComparisonResult.Equal.toString())
                    )
                    equalHead += genTriple(
                        empty,
                        compareResultPredicate,
                        value
                    )

                    // Правило для неэквивалентности
                    var notEqualHead = head0 + head1
                    notEqualHead += genNotEqualPrim(compiledArg0.value, compiledArg1.value)
                    // Инициализация value
                    notEqualHead += genTriple(
                        empty,
                        compareResultPredicate,
                        genVal(compiler.util.ComparisonResult.NotEqual.toString())
                    )
                    notEqualHead += genTriple(
                        empty,
                        compareResultPredicate,
                        value
                    )

                    // Добавляем в рзультат
                    heads.add(equalHead)
                    heads.add(notEqualHead)
                } else {
                    // Правило для эквивалентности
                    var equalHead = head0 + head1
                    equalHead += genEqualPrim(compiledArg0.value, compiledArg1.value)
                    // Инициализация value
                    equalHead += genTriple(
                        empty,
                        compareResultPredicate,
                        genVal(compiler.util.ComparisonResult.Equal.toString())
                    )
                    equalHead += genTriple(
                        empty,
                        compareResultPredicate,
                        value
                    )

                    // Правило для меньше
                    var lessHead = head0 + head1
                    lessHead += genLessThanPrim(compiledArg0.value, compiledArg1.value)
                    // Инициализация value
                    lessHead += genTriple(
                        empty,
                        compareResultPredicate,
                        genVal(compiler.util.ComparisonResult.Less.toString())
                    )
                    lessHead += genTriple(
                        empty,
                        compareResultPredicate,
                        value
                    )

                    // Правило для больше
                    var greaterHead = head0 + head1
                    greaterHead += genGreaterThanPrim(compiledArg0.value, compiledArg1.value)
                    // Инициализация value
                    greaterHead += genTriple(
                        empty,
                        compareResultPredicate,
                        genVal(compiler.util.ComparisonResult.Greater.toString())
                    )
                    greaterHead += genTriple(
                        empty,
                        compareResultPredicate,
                        value
                    )

                    // Добавляем в рзультат
                    heads.add(equalHead)
                    heads.add(lessHead)
                    heads.add(greaterHead)
                }
            }
        }

        return CompilationResult(value, heads, completedRules)
    }

    override fun clone(): Operator {
        val newArgs = ArrayList<Operator>()

        args.forEach { arg ->
            newArgs.add(arg.clone())
        }

        return Compare(newArgs)
    }

    override fun clone(newArgs: List<Operator>): Operator {
        return Compare(newArgs)
    }
}